e2e/k8s/k8s.go

```diff
@@ -0,0 +1,330 @@
+		// for us.
+		c.Server = "https://" + apiServerAddr
+	} else {
+		return nil, fmt.Errorf("generated kubeconfig missing context set")
+	}
+	restCfg, err := k8s_clientcmd.NewDefaultClientConfig(*kubeconfig, nil).ClientConfig()
+	if err != nil {
+		return nil, fmt.Errorf("could not generate k8s REST API config: %w", err)
+	}
+
+	return &Cluster{
+		runtime:    runtime,
+		k3dCluster: clusterConfig.Cluster,
+		restConfig: restCfg,
+		nginxAddr:  httpAddr,
+	}, nil
+}
+
+// injectCurrentDockerNetwork reconfigures config to join the Docker network of
+// the current container. Fails if the function is not being called from inside
+// of a Docker container.
+func injectCurrentDockerNetwork(ctx context.Context, config *k3d_config.SimpleConfig) error {
+	hostname, err := os.Hostname()
+	if err != nil {
+		return fmt.Errorf("could not get hostname: %w", err)
+	}
+
+	cli, err := k3d_docker.GetDockerClient()
+	if err != nil {
+		return fmt.Errorf("failed to get docker client: %w", err)
+	}
+	info, err := cli.ContainerInspect(ctx, hostname)
+	if err != nil {
+		return fmt.Errorf("failed to find current docker container: %w", err)
+	}
+
+	networks := make([]string, 0, len(info.NetworkSettings.Networks))
+	for nw := range info.NetworkSettings.Networks {
+		networks = append(networks, nw)
+	}
+	sort.Strings(networks)
+
+	if len(networks) == 0 {
+		return fmt.Errorf("no networks")
+	}
+	config.Network = networks[0]
+	return nil
+}
+
+func randomClusterName() string {
+	return "grafana-agent-e2e-" + rand.String(5)
+}
+
+func clusterInternalAddrs(ctx context.Context, cluster k3d_types.Cluster) (httpAddr, serverAddr string, err error) {
+	var lb, server *k3d_types.Node
+	for _, n := range cluster.Nodes {
+		switch n.Role {
+		case k3d_types.LoadBalancerRole:
+			if lb == nil {
+				lb = n
+			}
+		case k3d_types.ServerRole:
+			if server == nil {
+				server = n
+			}
+		}
+	}
+	if lb == nil {
+		return "", "", fmt.Errorf("no loadbalancer node")
+	} else if server == nil {
+		return "", "", fmt.Errorf("no server node")
+	}
+
+	cli, err := k3d_docker.GetDockerClient()
+	if err != nil {
+		return "", "", fmt.Errorf("failed to get docker client: %w", err)
+	}
+
+	lbInfo, err := cli.ContainerInspect(ctx, lb.Name)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to inspect loadbalancer: %w", err)
+	} else if len(lbInfo.NetworkSettings.Networks) == 0 {
+		return "", "", fmt.Errorf("no http address from loadbalancer")
+	} else {
+		for _, nw := range lbInfo.NetworkSettings.Networks {
+			httpAddr = fmt.Sprintf("%s:80", nw.IPAddress)
+			break
+		}
+	}
+
+	serverInfo, err := cli.ContainerInspect(ctx, server.Name)
+	if err != nil {
+		return "", "", fmt.Errorf("failed to inspect loadbalancer: %w", err)
+	} else if len(serverInfo.NetworkSettings.Networks) == 0 {
+		return "", "", fmt.Errorf("no API server address from worker")
+	} else {
+		for _, nw := range serverInfo.NetworkSettings.Networks {
+			serverAddr = fmt.Sprintf("%s:6443", nw.IPAddress)
+			break
+		}
```

Comment:
Will this cause an issue if there is more than one network? Would it be better to look up the network endpoint settings in this map by name? Since the iteration order of a map isn't guaranteed, it seems like this could possibly cause some intermittent issues that are hard to debug when several docker networks exist. Same for the lbInfo loop above.